File: leakcanary-analyzer/src/test/java/com/squareup/leakcanary/HeapAnalyzerTest.java
package com.squareup.leakcanary;

import com.squareup.haha.perflib.RootObj;
import com.squareup.haha.perflib.Snapshot;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

import static com.squareup.haha.perflib.RootType.NATIVE_STATIC;
import static com.squareup.haha.perflib.RootType.SYSTEM_CLASS;
import static com.squareup.leakcanary.TestUtil.NO_EXCLUDED_REFS;
import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;

@RunWith(JUnit4.class)
public class HeapAnalyzerTest {
  private static final List<RootObj> DUP_ROOTS =
          asList(new RootObj(SYSTEM_CLASS, 6L),
                  new RootObj(SYSTEM_CLASS, 5L),
                  new RootObj(SYSTEM_CLASS, 3L),
                  new RootObj(SYSTEM_CLASS, 5L),
                  new RootObj(NATIVE_STATIC, 3L));

  private HeapAnalyzer heapAnalyzer;

  @Before
  public void setUp() {
    heapAnalyzer = new HeapAnalyzer(NO_EXCLUDED_REFS);
  }

  @Test
  public void ensureUniqueRoots() {
    Snapshot snapshot = createSnapshot(DUP_ROOTS);

    heapAnalyzer.deduplicateGcRoots(snapshot);

    Collection<RootObj> uniqueRoots = snapshot.getGCRoots();
    assertThat(uniqueRoots).hasSize(4);

    List<Long> rootIds = new ArrayList<>();
    for (RootObj root : uniqueRoots) {
      rootIds.add(root.getId());
    }
    Collections.sort(rootIds);

    // 3 appears twice because even though two RootObjs have the same id, they're different types.
    assertThat(rootIds).containsExactly(3L, 3L, 5L, 6L);
  }

  private Snapshot createSnapshot(List<RootObj> gcRoots) {
    Snapshot snapshot = new Snapshot(null);
    for (RootObj root : gcRoots) {
      snapshot.addRoot(root);
    }
    return snapshot;
  }
}

File: leakcanary-analyzer/src/test/java/leakcanary/HeapAnalyzerTest.kt
package leakcanary

import com.squareup.haha.perflib.RootObj
import com.squareup.haha.perflib.RootType.NATIVE_STATIC
import com.squareup.haha.perflib.RootType.SYSTEM_CLASS
import com.squareup.haha.perflib.Snapshot
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.util.Arrays.asList

@RunWith(JUnit4::class)
class HeapAnalyzerTest {

  private lateinit var heapAnalyzer: HeapAnalyzer

  @Before
  fun setUp() {
    heapAnalyzer = HeapAnalyzer(
        NO_EXCLUDED_REFS, AnalyzerProgressListener.NONE,
        emptyList()
    )
  }

  @Test
  fun ensureUniqueRoots() {
    val snapshot = createSnapshot(DUP_ROOTS)

    heapAnalyzer.deduplicateGcRoots(snapshot)

    val uniqueRoots = snapshot.gcRoots
    assertThat(uniqueRoots).hasSize(4)

    val rootIds = mutableListOf<Long>()
    uniqueRoots.forEach { root ->
      rootIds.add(root.id)
    }
    rootIds.sort()

    // 3 appears twice because even though two RootObjs have the same id, they're different types.
    assertThat(rootIds).containsExactly(3L, 3L, 5L, 6L)
  }

  private fun createSnapshot(gcRoots: List<RootObj>): Snapshot {
    val snapshot = Snapshot(null)
    for (root in gcRoots) {
      snapshot.addRoot(root)
    }
    return snapshot
  }

  companion object {
    private val DUP_ROOTS = asList<RootObj>(
        RootObj(SYSTEM_CLASS, 6L),
        RootObj(SYSTEM_CLASS, 5L),
        RootObj(SYSTEM_CLASS, 3L),
        RootObj(SYSTEM_CLASS, 5L),
        RootObj(NATIVE_STATIC, 3L)
    )
  }
}
